pkg/controller/grouping/group_entity_index.go

```diff
@@ -0,0 +1,728 @@
+	labels labels.Set
+	// The Namespace of the entities that share the labelItem.
+	namespace string
+	// The type of the entities that share the labelItem.
+	entityType entityType
+	// The keys of the entityItems that share the labelItem.
+	entityItemKeys sets.String
+	// The keys of the selectorItems that match the labelItem.
+	selectorItemKeys sets.String
+}
+
+// groupItem contains a group's metadata and its selector.
+type groupItem struct {
+	// The type of the group.
+	groupType string
+	// The name of the group. It must be unique within its own type.
+	name string
+	// The selector of the group.
+	selector *types.GroupSelector
+	// selectorItemKey is the key of the selectorItem that the groupItem is associated with.
+	// groupItems will be associated with the same selectorItem if they have same selector.
+	selectorItemKey string
+}
+
+// selectorItem represents an individual label selector. It's the actual object that will be matched with label sets.
+// Groups having same label selector will share a selectorItem.
+type selectorItem struct {
+	// The label selector that will be used for matching.
+	selector *types.GroupSelector
+	// The keys of the groupItems that share the selectorItem.
+	groupItemKeys sets.String
+	// The keys of the labelItems that match the selectorItem.
+	labelItemKeys sets.String
+}
+
+var _ Interface = &GroupEntityIndex{}
+
+// GroupEntityIndex implements Interface.
+//
+// It abstracts label set and label selector from entities and groups and does actual matching against the formers to
+// avoid redundant calculation given that most entities (Pod or ExternalEntity) actually share labels. For example, Pods
+// that managed by a deployment controller have same labels.
+//
+// It maintains indexes between label set and label selector so that querying label sets that match a label selector and
+// reversed queries can be performed with a constant time complexity. Indirectly, querying entities that match a group
+// and reversed queries can be performed with same complexity.
+//
+// The relationship of the four items are like below:
+// entityItem <===> labelItem <===> selectorItem <===> groupItem
+type GroupEntityIndex struct {
+	lock sync.RWMutex
+
+	// entityItems stores all entityItems.
+	entityItems map[string]*entityItem
+
+	// labelItems stores all labelItems.
+	labelItems map[string]*labelItem
+	// labelItemIndex is nested map from entityType to Namespace to keys of labelItems.
+	// It's used to filter potential labelItems when matching a Namespace scoped selectorItem.
+	labelItemIndex map[entityType]map[string]sets.String
+
+	// groupItems stores all groupItems.
+	groupItems map[string]*groupItem
+
+	// selectorItems stores all selectorItems.
+	selectorItems map[string]*selectorItem
+	// selectorItemIndex is nested map from entityType to Namespace to keys of selectorItems.
+	// It's used to filter potential selectorItems when matching an labelItem.
+	selectorItemIndex map[entityType]map[string]sets.String
```

Comment:
Is it easier to understand if we use a separate map[entityType]sets.String for cluster scope selectors?

Reply:
Comments definitely help. But I thought your usage pattern is always to check the selector is cluster scope or not, and if so look up with "". Then it does not help simplify the code to reuse a single map? Maybe I missed something.

Reply:
They are treated in same way when maintaing the selectorItemIndex:
https://github.com/vmware-tanzu/antrea/blob/413aad795ae599ef061811a552541f27358b32ed/pkg/controller/grouping/group_entity_index.go#L480
https://github.com/vmware-tanzu/antrea/blob/413aad795ae599ef061811a552541f27358b32ed/pkg/controller/grouping/group_entity_index.go#L505-L509

Reply:
Got it. Fine with "" (do you think a constant helps).

Reply:
Updated to constant emptyNamespace